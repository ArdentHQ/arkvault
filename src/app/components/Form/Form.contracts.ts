import React from "react";
import { SubmitHandler, UseFormMethods } from "react-hook-form";
import { FieldValues } from "react-hook-form/dist/types/fields";

export type FormProperties<T extends FieldValues = any> = {
	onSubmit?: SubmitHandler<T>;
	context: UseFormMethods<T>;
} & Omit<React.FormHTMLAttributes<Element>, "onSubmit">;
