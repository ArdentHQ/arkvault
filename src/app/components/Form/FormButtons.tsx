import React, { useEffect } from "react";
import tw, { styled, css } from "twin.macro";
import { useNavigationContext } from "@/app/contexts";

const FormButtonsWrapper = styled.div<{
	showMobileNavigation?: boolean;
}>`
	${({ showMobileNavigation }) => showMobileNavigation && tw`sm:mb-0`};
	${({ showMobileNavigation }) =>
		showMobileNavigation &&
		css`
			@media (max-width: 639px) {
				margin-bottom: 3.5rem;
				@supports (margin-bottom: env(safe-area-inset-bottom)) {
					margin-bottom: calc(env(safe-area-inset-bottom) + 3.5rem);
				}
			}
		`};
	${({ showMobileNavigation }) =>
		!showMobileNavigation &&
		css`
			@media (max-width: 639px) {
				@supports (padding-bottom: env(safe-area-inset-bottom)) {
					padding-bottom: calc(env(safe-area-inset-bottom) + 0.75rem);
				}
			}
		`};
	${tw`flex fixed bg-theme-background dark:bg-black sm:dark:bg-transparent inset-x-0 bottom-0 px-8 py-3 gap-3 shadow-footer-smooth dark:shadow-footer-smooth-dark`}
	${tw`sm:(relative inset-auto p-0 mt-8 justify-end shadow-none dark:shadow-none)`}
	${css`
		& > button {
			${tw`flex-1 sm:flex-none`}
		}
	`}
`;

const FormButtons: React.FC<{
	children: React.ReactNode;
}> = ({ children }) => {
	const { showMobileNavigation, setHasFixedFormButtons } = useNavigationContext();

	useEffect(() => {
		setHasFixedFormButtons(true);
		return () => {
			setHasFixedFormButtons(false);
		};
	});

	return <FormButtonsWrapper showMobileNavigation={showMobileNavigation}>{children}</FormButtonsWrapper>;
};

export { FormButtons, FormButtonsWrapper };
