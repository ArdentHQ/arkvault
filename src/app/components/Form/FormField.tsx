import { get } from "@ardenthq/sdk-helpers";
import React from "react";
import { useFormContext } from "react-hook-form";

import { FormFieldProvider } from "./useFormField";
import { twMerge } from "tailwind-merge";
import cn from "classnames";

type FormFieldProperties = {
	name: string;
	disableHover?: boolean;
} & React.FieldsetHTMLAttributes<any>;

export const FormFieldStyled = ({
	isInvalid,
	disableHover,
	...props
}: { isInvalid: boolean; disableHover: boolean } & React.FieldsetHTMLAttributes<HTMLFieldSetElement>) => (
	<fieldset
		{...props}
		className={twMerge(
			cn({
				"[&:focus-within_.FormLabel]:text-theme-danger-500": isInvalid,
				"[&:focus-within_.FormLabel]:text-theme-primary-600 dark:[&:focus-within_.FormLabel]:text-theme-primary-500":
					!isInvalid && !disableHover,
				"[&:focus-within_.FormLabel]:text-theme-secondary-text": !isInvalid && disableHover,
				"[&>.FormLabel]:text-theme-danger-500": isInvalid,
			}),
			props.className,
		)}
	/>
);

export const FormField: React.FC<FormFieldProperties> = ({ name, disableHover = false, ...properties }) => {
	const FormProvider = useFormContext();
	const { isInvalid, errorMessage } = React.useMemo(() => {
		const error: { message: string } | undefined = get(FormProvider?.errors, name);

		return {
			errorMessage: error?.message,
			isInvalid: !!error,
		};
	}, [FormProvider, name]);

	return (
		<FormFieldStyled
			isInvalid={isInvalid}
			className="flex min-w-0 flex-col"
			disableHover={disableHover}
			{...properties}
		>
			<FormFieldProvider value={{ errorMessage, isInvalid, name }}>
				<>{properties.children}</>
			</FormFieldProvider>
		</FormFieldStyled>
	);
};
