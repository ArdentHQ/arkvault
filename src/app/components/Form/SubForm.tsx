import React from "react";
import { twMerge } from "tailwind-merge";
import cn from "classnames";

const SubFormWrapper = ({
	noPadding,
	noBackground,
	...props
}: React.HTMLAttributes<HTMLDivElement> & { noBackground?: boolean; noPadding?: boolean }) => (
	<div
		{...props}
		className={twMerge(
			"space-y-4 rounded-lg",
			cn({
				"-mx-4 p-4": !noPadding,
				"bg-theme-secondary-background": !noBackground,
			}),
			props.className,
		)}
	/>
);

export const SubForm = ({
	className,
	children,
	noBackground,
	noPadding,
}: {
	className?: string;
	children: React.ReactNode;
	noBackground?: boolean;
	noPadding?: boolean;
}) => (
	<SubFormWrapper className={className} noBackground={noBackground} noPadding={noPadding}>
		{children}
	</SubFormWrapper>
);
