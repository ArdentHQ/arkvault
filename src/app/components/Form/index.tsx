export * from "./Form";
export * from "./Form.contracts";
export * from "./FormButtons";
export * from "./FormField";
export * from "./FormHelperText";
export * from "./FormLabel";
export * from "./SubForm";
