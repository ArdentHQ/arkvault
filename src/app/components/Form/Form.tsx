import React from "react";
import { FormProvider } from "react-hook-form";

import { FormProperties } from "@/app/components/Form/Form.contracts";

export const Form = React.forwardRef<HTMLFormElement, FormProperties>(
	({ children, context, onSubmit, ...properties }, reference) => (
		<FormProvider {...context}>
			<form
				data-testid="Form"
				ref={reference}
				className="space-y-5"
				onSubmit={onSubmit ? context.handleSubmit(onSubmit) : (event) => event.preventDefault()}
				{...properties}
			>
				{children}
			</form>
		</FormProvider>
	),
);

Form.displayName = "Form";
