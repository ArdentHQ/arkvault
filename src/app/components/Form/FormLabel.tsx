import React from "react";
import { useTranslation } from "react-i18next";

import { useFormField } from "./useFormField";
import { Tooltip } from "@/app/components/Tooltip";
import cn from "classnames";

type FormLabelProperties = {
	label?: string;
	optional?: boolean;
	disableHints?: boolean;
} & React.LabelHTMLAttributes<any>;

export function FormLabel(properties: FormLabelProperties) {
	const fieldContext = useFormField();

	const labelProperties = { ...properties };

	for (const property of ["label", "optional", "disableHints"]) {
		// @ts-ignore
		delete labelProperties[property];
	}

	const { t } = useTranslation();

	return (
		<label
			data-testid="FormLabel"
			className={cn("mb-2 flex text-sm font-semibold leading-[17px] text-theme-secondary-text transition-colors duration-100 hover:!text-theme-primary-600", {"FormLabel": !properties.disableHints})}
			htmlFor={fieldContext?.name}
			{...labelProperties}
		>
			<>
				{properties.label || properties.children}

				{properties.optional && (
					<Tooltip content={t("COMMON.VALIDATION.OPTIONAL")}>
						<span
							data-testid="FormLabel__optional"
							className="ml-1 text-theme-secondary-500 dark:text-theme-secondary-700"
						>
							{t("COMMON.OPTIONAL")}
						</span>
					</Tooltip>
				)}
			</>
		</label>
	);
}
