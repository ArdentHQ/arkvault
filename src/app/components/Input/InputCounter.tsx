import cn from "classnames";
import React, { useState } from "react";

import { useFormField } from "@/app/components/Form/useFormField";
import { Input } from "@/app/components/Input";

type Properties = {
	maxLength?: number;
	maxLengthLabel?: string;
	defaultValue?: string;
} & React.InputHTMLAttributes<any>;

export const InputCounter = React.forwardRef<HTMLInputElement, Properties>((properties: Properties, reference) => {
	const fieldContext = useFormField();
	const [length, setLength] = useState(properties.defaultValue?.length || 0);

	const handleChange = (event: React.ChangeEvent<HTMLInputElement>) => {
		setLength(event.target.value.length);
		properties.onChange?.(event);
	};

	return (
		<Input
			data-testid="InputCounter__input"
			ref={reference}
			{...properties}
			onChange={handleChange}
			addons={{
				end: {
					content: (
						<span
							data-testid="InputCounter__counter"
							className={cn("text-sm font-semibold", {
								"text-theme-secondary-500 dark:text-theme-secondary-700": !fieldContext?.isInvalid,
							})}
						>
							{length}/{properties.maxLengthLabel}
						</span>
					),
				},
			}}
		/>
	);
});

InputCounter.displayName = "InputCounter";
