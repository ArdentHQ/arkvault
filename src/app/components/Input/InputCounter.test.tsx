import userEvent from "@testing-library/user-event";
import React from "react";

import { InputCounter } from "./InputCounter";
import { FormFieldProvider } from "@/app/components/Form/useFormField";
import { render, screen } from "@/utils/testing-library";

describe("InputCounter", () => {
	it("should render", () => {
		const { asFragment } = render(<InputCounter maxLength={10} maxLengthLabel="10" />);

		expect(screen.getByTestId("InputCounter__counter")).toHaveTextContent("0/10");
		expect(screen.getByTestId("InputCounter__input")).toHaveAttribute("maxLength", "10");
		expect(asFragment()).toMatchSnapshot();
	});

	it("should render with default value", () => {
		render(<InputCounter maxLength={10} maxLengthLabel="10" defaultValue="Hello" />);

		expect(screen.getByTestId("InputCounter__counter")).toHaveTextContent("5/10");
	});

	it("should update the length when changing the value", () => {
		render(<InputCounter maxLength={10} maxLengthLabel="10" />);

		userEvent.type(screen.getByTestId("InputCounter__input"), "Test");

		expect(screen.getByTestId("InputCounter__counter")).toHaveTextContent("4/10");
	});

	it("should render with invalid state", () => {
		const { container } = render(
			<FormFieldProvider value={{ isInvalid: true, name: "vendorField" }}>
				<InputCounter maxLength={10} maxLengthLabel="10" />
			</FormFieldProvider>,
		);

		expect(container).toMatchSnapshot();
	});
});
