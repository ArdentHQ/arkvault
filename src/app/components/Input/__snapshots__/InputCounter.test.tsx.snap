// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`InputCounter > should render 1`] = `
<DocumentFragment>
  <div
    class="css-1psxlcd"
  >
    <div
      class="relative flex h-full flex-1"
    >
      <input
        autocomplete="off"
        class="no-ligatures w-full border-none css-1nupw50"
        data-testid="InputCounter__input"
        maxlength="10"
        type="text"
        value=""
      />
    </div>
    <div
      class="flex items-center space-x-3 divide-x divide-theme-secondary-300 dark:divide-theme-secondary-800 text-theme-primary-300 dark:text-theme-secondary-600"
      data-testid="Input__addon-end"
    >
      <div
        class=""
      >
        <span
          class="text-sm font-semibold text-theme-secondary-500 dark:text-theme-secondary-700"
          data-testid="InputCounter__counter"
        >
          0/10
        </span>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`InputCounter > should render with invalid state 1`] = `
<div>
  <div
    class="css-px3dod"
  >
    <div
      class="relative flex h-full flex-1"
    >
      <input
        aria-invalid="true"
        autocomplete="off"
        class="no-ligatures w-full border-none css-1nupw50"
        data-testid="InputCounter__input"
        maxlength="10"
        name="vendorField"
        type="text"
        value=""
      />
    </div>
    <div
      class="flex items-center space-x-3 divide-x divide-theme-secondary-300 dark:divide-theme-secondary-800 text-theme-danger-500"
      data-testid="Input__addon-end"
    >
      <span
        data-testid="Input__error"
      >
        <div
          class="text-theme-danger-500 css-1i4b0eq"
          height="20"
          width="20"
        >
          <svg
            id="circle-exclamation-mark"
            viewBox="0 0 20 20"
            x="0"
            xml:space="preserve"
            xmlns="http://www.w3.org/2000/svg"
            y="0"
          >
            <g
              fill="none"
              stroke="currentColor"
              stroke-width="2"
            >
              <circle
                cx="10"
                cy="10"
                r="9"
                stroke-miterlimit="10"
              />
              <path
                d="M10 10.5v-5"
                stroke-linecap="round"
                stroke-miterlimit="10"
              />
              <path
                d="M10 14h0"
                stroke-linecap="round"
                stroke-linejoin="round"
              />
            </g>
          </svg>
        </div>
      </span>
      <div
        class="pl-3"
      >
        <span
          class="text-sm font-semibold"
          data-testid="InputCounter__counter"
        >
          0
          /
          10
        </span>
      </div>
    </div>
  </div>
</div>
`;
